pandas/tests/dtypes/cast/test_promote.py

```diff
@@ -0,0 +1,624 @@
+# -*- coding: utf-8 -*-
+
+"""
+These test the method maybe_promote from core/dtypes/cast.py
+"""
+
+import datetime
+
+import numpy as np
+import pytest
+
+from pandas._libs.tslibs import NaT, iNaT
+from pandas.compat import is_platform_windows
+
+from pandas.core.dtypes.cast import maybe_promote
+from pandas.core.dtypes.common import (
+    is_complex_dtype, is_datetime64_dtype, is_datetime_or_timedelta_dtype,
+    is_float_dtype, is_integer_dtype, is_object_dtype, is_scalar,
+    is_string_dtype, is_timedelta64_dtype)
+from pandas.core.dtypes.dtypes import DatetimeTZDtype
+
+import pandas as pd
+
+
+def _check_promote(dtype, fill_value, boxed, box_dtype, expected_dtype,
+                   exp_val_for_scalar=None, exp_val_for_array=None):
+    assert is_scalar(fill_value)
+
+    if boxed:
+        fill_array = np.array([fill_value], dtype=box_dtype)
+        result_dtype, result_fill_value = maybe_promote(dtype, fill_array)
+        expected_fill_value = exp_val_for_array
+    else:
+        result_dtype, result_fill_value = maybe_promote(dtype, fill_value)
+        expected_fill_value = exp_val_for_scalar
+
+    # try/except as numpy dtypes (i.e. if result_dtype is np.object_) do not
+    # know some expected dtypes like DatetimeTZDtype, and hence raise TypeError
+    try:
+        assert result_dtype == expected_dtype
+    except TypeError:
+        assert expected_dtype == result_dtype
+
```

Comment:
handle null values in a separate clause to make this less compound

Reply:
OK